lib/promscrape/config.go

```diff
@@ -102,6 +102,52 @@ func (cfg *Config) mustStart() {
 	logger.Infof("started service discovery routines in %.3f seconds", time.Since(startTime).Seconds())
 }
 
+func (cfg *Config) mustStartRestart(prevCfg *Config) {
+	startTime := time.Now()
+	logger.Infof("restarting service discovery routines...")
+
+	prevScrapeCfgByName := make(map[string]*ScrapeConfig, len(prevCfg.ScrapeConfigs))
+	for i := range prevCfg.ScrapeConfigs {
+		sc := &prevCfg.ScrapeConfigs[i]
+		prevScrapeCfgByName[sc.JobName] = sc
+	}
+
+	var toStart, toStop []*ScrapeConfig
+	currentJobNames := make(map[string]struct{}, len(cfg.ScrapeConfigs))
+	for i := range cfg.ScrapeConfigs {
+		sc := cfg.ScrapeConfigs[i]
+		currentJobNames[sc.JobName] = struct{}{}
+		prevSC := prevScrapeCfgByName[sc.JobName]
+
+		if prevSC != nil && prevSC.checksum == sc.checksum {
+			// config checksum the same, no need to restart it
+			// replace reference with previous job, it allows to stop it properly later.
+			cfg.ScrapeConfigs[i] = *prevSC
+			continue
+		}
+		if prevSC != nil {
+			toStop = append(toStop, prevSC)
+		}
+		toStart = append(toStart, &sc)
+	}
+	// add toStop jobs, that was not found at current configuration
+	for i := range prevCfg.ScrapeConfigs {
+		prevSC := &prevCfg.ScrapeConfigs[i]
+		if _, ok := currentJobNames[prevSC.JobName]; !ok {
+			toStop = append(toStop, prevSC)
+		}
+	}
+	for _, sc := range toStop {
+		sc.mustStop()
+	}
+	for _, sc := range toStart {
+		sc.mustStart(cfg.baseDir)
+	}
```

Comment:
Does it make sense to parallel this work? Or parallel stopping/starting only?

Reply:
I thought about it. And i think, that it's not needed. start and stop must be pretty fast. And it could paralleled at service_discovery level.